### PR TITLE
Add event processor functionality

### PR DESCRIPTION
**Pull request by caaavik-msft**
This addresses #2054 which is a feature request to have hooks which can enable other tools that are invoking BDN to display progress rather than having to parse live console output or to wait for all the benchmarks to complete to have the exporters run. 

A full proposal and explanation of the events/hooks in this PR are covered in this comment: https://github.com/dotnet/BenchmarkDotNet/issues/2054#issuecomment-1202229339

There are potentially many more hooks that could be added like events for when a benchmark run enters a different stage, but they would likely need to be added as custom events sent through the Broker and supported by all the different executors/toolchains, which I didn't want to add at this stage. I'm open as well to modifying or adding extra events before merging this in to see what people think would be useful.

I've also included a `EventHandlerBase` class which is a base class that implements all the hooks as virtual methods, so that implementing classes only need to override the hooks they care about.

### CODE REVIEW

**Comment by cincuranet** [comment]
Any reason why `CompositeEventHandler` does not derive from `EventHandlerBase`?

**Comment by caaavik-msft** [comment]
> Any reason why `CompositeEventHandler` does not derive from `EventHandlerBase`?

`EventHandlerBase` is just an implementation of the interface with virtual methods so that custom event handlers don't have to implement every hook and can just implement the hooks they care about. The `CompositeEventHandler` composes multiple event handlers into one, so by nature we want to ensure that every hook is implemented and handled, if we used the EventHandlerBase we may forget to update `CompositeEventHandler` if we add a new hook.

File: src/BenchmarkDotNet/EventHandlers/EventHandlerBase.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventHandlers/EventHandlerBase.cs:10]: I know that all existing BDN config switches expose interfaces, but in this case I wonder whether we should just expose an `abstract` class with all the methods being `virtual`. Benefits:
- to sign up for a single event, the users would need to override only one method
- if we wanted to introduce a new event in the future, we would not break the existing users, as their code would still compile.

What do you think?

File: src/BenchmarkDotNet/EventHandlers/IEventHandler.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventHandlers/IEventHandler.cs:10]: `EventHandler` has a very specific meaning in C#: https://learn.microsoft.com/en-us/dotnet/api/system.eventhandler?view=net-7.0

Adding an `I` in front of it may suggest that it's an interface that `EventHandler` implements.

I believe that we should find a better name. I don't have any good suggestions, but perhaps @bartonjs could throw us a hand here?

File: src/BenchmarkDotNet/EventHandlers/IEventHandler.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventHandlers/IEventHandler.cs:13]: To me the **Handle**UnsupportedBenchmark name suggests that you can actually do something with the unsuported benchmark (like stop benchmarking or continue). Since it's not the case here, I think we should use the `On$Abc` rather than `Handle$Abc` naming convention.

File: src/BenchmarkDotNet/EventHandlers/IEventHandler.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventHandlers/IEventHandler.cs:16]: we have events for build start and failure, what about successful completion?

File: src/BenchmarkDotNet/EventHandlers/IEventHandler.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventHandlers/IEventHandler.cs:19]: the naming is inconsistent: Start&Stop vs Run&Completed. We should choose one schema and stick to it.

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:73]: I am not sure whether we should be allowing to have the event handlers only for selected benchmarks.

```suggestion
                    eventHandler.HandleValidationError(validationError);
```

@caaavik-msft what do you think? do you see any real life usage examples of that?

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:212]: could you please pass an instance of composite handler from the calling method here (we already do that for logger)?

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:378]: you have achieved exactly the same effect, but with more code. Did you intend to raise some events here?

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:386]: this variable is not used, did you mean to use it?

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:608]: `OnValidaitonError` seems to be a better name (not every validation error is a hard stop)

File: tests/BenchmarkDotNet.IntegrationTests/EventHandlerTests.cs
Comment (adamsitnik) [inline tests/BenchmarkDotNet.IntegrationTests/EventHandlerTests.cs:160]: thank you for using `Dry` job here 👍

File: tests/BenchmarkDotNet.IntegrationTests/EventHandlerTests.cs
Comment (adamsitnik) [inline tests/BenchmarkDotNet.IntegrationTests/EventHandlerTests.cs:298]: big thanks for adding plenty of tests! 🥇

**Review by adamsitnik — COMMENTED** [review]
Overall it looks good, but let's iterate on simplifying it a bit and unifying the naming.

Thank you for your contribution @caaavik-msft !

File: src/BenchmarkDotNet/EventHandlers/EventHandlerBase.cs
Comment (bartonjs) [reply src/BenchmarkDotNet/EventHandlers/EventHandlerBase.cs:10]: I was about to say something very similar.  The fact that this is "one type to handle all kinds of events" means an interface is not suitable.

File: src/BenchmarkDotNet/EventHandlers/IEventHandler.cs
Comment (bartonjs) [reply src/BenchmarkDotNet/EventHandlers/IEventHandler.cs:10]: Event...Processor?  Event...Broker?

Comment (timcassell) [reply src/BenchmarkDotNet/EventHandlers/IEventHandler.cs:10]: EventListener?

Comment (caaavik-msft) [reply src/BenchmarkDotNet/EventHandlers/IEventHandler.cs:10]: what about `BenchmarkEventHandler`? I'm going to use this for now but if we want to go with something else then am happy to change it again

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (caaavik-msft) [reply src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:73]: The only case I can think of would be for the BenchmarkDotNet.Samples project where we might want to demonstrate a custom event handler but not apply it to every benchmark that is being run. But I suppose even for that case it is probably acceptable since usually only one type is run at a time when using the samples.

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (caaavik-msft) [reply src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs]: I have removed `OnUnsupportedBenchmark` since they surface in `OnValidationError` anyways

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (timcassell) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:98]: Why not just do `OnBuildComplete` so the listener can check for build success?

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (timcassell) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:87]: It might be handy to pass in how many partitions are being built.

File: src/BenchmarkDotNet/EventHandlers/CompositeBenchmarkEventHandler.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventHandlers/CompositeBenchmarkEventHandler.cs:10]: this type does not need to be `public` 

```suggestion
    internal sealed class CompositeBenchmarkEventHandler : BenchmarkEventHandlerBase
```

File: src/BenchmarkDotNet/EventHandlers/BenchmarkEventHandlerBase.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventHandlers/BenchmarkEventHandlerBase.cs:10]: this is very subjective, but `EventProcessor` suggested by @bartonjs sounds best to me.

```suggestion
    public abstract class EventProcessor
```

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:678]: this logic can be moved the the ctor of the composite event handler/processor

**Review by adamsitnik — COMMENTED** [review]
We are almost there, thank you for addressing the feedback @caaavik-msft !

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (timcassell) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:392]: It looks like this is being called concurrently, which necessitates synchronization. Perhaps it would be better to take the parallel query and force it back to sequential to invoke on the same thread?

```cs
buildResults = buildPartitions
    .AsParallel()
    .Select(buildPartition => (buildPartition, buildResult: Build(buildPartition, rootArtifactsFolderPath, buildLogger)))
    .AsSequential()
    .ToDictionary(...)
```

Comment (caaavik-msft) [reply src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs:392]: Nice suggestion, I did not know that `AsSequential` existed. Adding that in now.

**Review by timcassell — APPROVED** [review]

**Comment by caaavik-msft** [comment]
@adamsitnik are we good to merge this in now?

File: src/BenchmarkDotNet/EventProcessors/CompositeEventProcessor.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventProcessors/CompositeEventProcessor.cs:13]: We should prefer concrete types over abstractions for private fields (https://github.com/dotnet/roslyn-analyzers/pull/6370)

```suggestion
        private readonly HashSet<EventProcessorBase> eventProcessors;
```

File: src/BenchmarkDotNet/EventProcessors/CompositeEventProcessor.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventProcessors/CompositeEventProcessor.cs:28]: it seems that this ctor is not used anywhere

```suggestion
```

File: src/BenchmarkDotNet/EventProcessors/EventProcessorBase.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventProcessors/EventProcessorBase.cs:17]: what is the difference between `OnBuildComplete` and `OnEndBuildStage`? The first is for one partition and the latter is for all of them?

File: src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Running/BenchmarkRunnerClean.cs]: thank you for adding the comment 👍

File: src/BenchmarkDotNet/EventProcessors/EventProcessorBase.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/EventProcessors/EventProcessorBase.cs:10]: in the past we used the `Base` suffix for abstract types, but from what I've learned from API Reviews it's an anti-pattern

```suggestion
    public abstract class EventProcessor
```

**Review by adamsitnik — COMMENTED** [review]
LGTM, but please rename the type before I hit the merge button.

Thank you @caaavik-msft !

File: src/BenchmarkDotNet/EventProcessors/EventProcessorBase.cs
Comment (caaavik-msft) [reply src/BenchmarkDotNet/EventProcessors/EventProcessorBase.cs:17]: Yes, that is the difference between them.

File: src/BenchmarkDotNet/Configs/ConfigExtensions.cs
Comment (adamsitnik) [inline src/BenchmarkDotNet/Configs/ConfigExtensions.cs:137]: this should fix the build

```suggestion
    }
}
```

File: src/BenchmarkDotNet/Configs/ConfigExtensions.cs
Comment (mawosoft) [reply src/BenchmarkDotNet/Configs/ConfigExtensions.cs]: @adamsitnik , @caaavik-msft

> nit: it's unlikely that the users will need more than one instance

While this may be true, this makes it the only one of the `Add...` methods that doesn't accept a params array. Personally, I would prefer the API consistency.

File: src/BenchmarkDotNet/Configs/ManualConfig.cs
Comment (mawosoft) [inline src/BenchmarkDotNet/Configs/ManualConfig.cs:227]: @adamsitnik, @caaavik-msft

> nit: it's unlikely that the users will need more than one instance

While this may be true, this makes it the only one of the `Add...` methods that doesn't accept a params array. Personally, I would prefer the API consistency.

**Comment by mawosoft** [comment]
Also, `ManualConfig.Add(IConfig config)` and, by extension, `ManualConfig.Union(IConfig globalConfig, IConfig localConfig)` and `ManualConfig.Create(IConfig config)` don't copy the event processor(s) from the source.

**Comment by timcassell** [comment]
@caaavik-msft Can you address @mawosoft's comment in another PR (merging configs)?

**Pull request merged**
